pkg/innerring/innerring.go

```diff
@@ -352,15 +346,88 @@ func New(ctx context.Context, log *logger.Logger, cfg *viper.Viper, errChan chan
 	morphChain := &chainParams{
 		log:  log,
 		cfg:  cfg,
-		key:  server.key,
 		name: morphPrefix,
 		from: fromSideChainBlock,
 	}
 
+	const walletPathKey = "wallet.path"
+	if !cfg.IsSet(walletPathKey) {
+		return nil, fmt.Errorf("file path to the node Neo wallet is not configured '%s'", walletPathKey)
+	}
+
+	walletPath := cfg.GetString(walletPathKey)
+	walletPass := cfg.GetString("wallet.password")
+
 	// create morph client
-	server.morphClient, err = createClient(ctx, morphChain, errChan)
-	if err != nil {
-		return nil, err
+	if len(cfg.GetStringSlice("morph.endpoint.client")) == 0 {
+		// go on a local blockchain
+		cfgBlockchain, err := parseBlockchainConfig(cfg, log)
+		if err != nil {
+			return nil, fmt.Errorf("invalid blockchain configuration: %w", err)
+		}
+
+		cfgBlockchain.Wallet.Path = walletPath
+		cfgBlockchain.Wallet.Password = walletPass
+		cfgBlockchain.ErrorListener = errChan
+
+		server.bc, err = blockchain.New(cfgBlockchain)
+		if err != nil {
+			return nil, fmt.Errorf("init internal blockchain: %w", err)
+		}
+
+		// don't move this code area within current function
+		{
+			log.Info("running blockchain...")
+			// we run node here instead of Server.Start because some Server components
+			// require blockchain to be run in the current function. Later it would be nice
+			// to separate the stages of construction and launch
+			err = server.bc.Run()
+			if err != nil {
+				return nil, fmt.Errorf("run internal blockchain: %w", err)
+			}
+
+			defer func() {
+				if err != nil {
+					server.bc.Stop()
+				}
+			}()
+		}
+
+		wsClient, err := server.bc.BuildWSClient(ctx)
```

Comment:
We build WSClient here (immediately after blockchain init) and then use it everywhere. This made me wonder about `StartWhenSynchronized` option and I think we have a problem with it.

The way it was is:
 * NeoGo starts independently and synchronizes with the network
 * NeoFS IR starts independently and tries to connect to NeoGo
 * if NeoGo is not yet in sync, the connection won't succeed, IR will fail init, but it's likely to be restarted
 * so effectively it'll retry again and again until it can connect, which happens when NeoGo is in sync with the network (`StartWhenSynchronized`)
 * everything works smoothly, IR only sees current state and events

Now we've got this:
 * internal blockchain starts, we build WSClient on top of its RPC server and it's always there
 * if the node is in sync, it's OK
 * if it's not, we have two problems:
   - internal RPC server's event handler loop is not running, but subscriptions can happen via internal client -> deadlock on event processing
   - potentially we're running as if `StartWhenSynchronized` is false, old events may be received by IR causing it to send some transactions which may then lead to notary balance drain

Options to solve it:
 * failing to create a client (requires NeoGo change), which is problematic as it will lead to IR restarts and not allow blockchain to sync, but can an option if we're to handle this particular error here
 * creating a client, but failing requests to it (requires NeoGo change), I'm not sure how this will be handled by IR (it'll try to reconnect, but get the error again?)
 * checking the status in IR itself, maybe the easiest to implement now (can be done in NeoFS itself), we have https://pkg.go.dev/github.com/nspcc-dev/neo-go/pkg/network#Server.IsInSync and it can be checked until it's `true`
 * add IR to the network server via `AddService` and let it be started when it's appropriate to do so

Thoughts, opinions?

Reply:
> add IR to the network server via AddService

Overall it looks better to me, but once we have #2294 it's not really easy to implement right now.

In current architecture IR depends on NeoGo node, so IR should start only after NeoGo is ready-to-go. I'd start with any working NeoGo-way to reach sync state (now it is vital for IR) and block `internal.Blockchain.Run` method until this state. This will keep the current component hierarchy up and running.

Speaking of handling events from the past: I would come up with some generic way for the IR (e.g. for alternative implementation) to avoid handling them. Maybe sync-await will fix it or at least reduce to a minimum?

Reply:
> add IR to the network server via AddService

In perfect world it's my favourite option. And doing this will require move all IR/RPC client/etc initialisation staff to the `service.Start()` callback which will make our code much cleaner.

But since this PR is just an initial implementation, I vote for the third option (polling waiting for `server.IsInSync()` to become `true`), and after that we can safely use subscriptions and to be sure that all notifications are up-to-date. First two options are harder to handle and look not natural to the neo-go itself. It's not clear to me why the client should fail the construction/requists just because the server is out of sync.

Reply:
Talking about async/await:
1. as an alternative to polling, I see a special  `network.Service` as a working option, the `Start` of which will mean the onset of synchronization (taking into account the always-on `StartWhenSynchronized` option). This seems a bit complicated, but polling is always less transparent approach overall to me.
2. if we choose polling approach: what value is better to use as a time period?

Reply:
2 --- block time, as usual. I think it's OK for us to poll for now, blockchain threads will do their job in the background and then IR will kick in. Let's leave more complex approaches to #2294.

Reply:
Done